diff --git a/external-module-compat-comm.h b/external-module-compat-comm.h
index 34fb320..6dd2074 100644
--- a/external-module-compat-comm.h
+++ b/external-module-compat-comm.h
@@ -1214,7 +1214,7 @@ static inline bool pci_check_and_unmask_intx(struct pci_dev *dev)
 #define IRQ_WAKE_THREAD		IRQ_NONE	/* will never be used */
 #endif
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(3,4,0)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,2,0)
 struct x86_cpu_id { };
 #define X86_FEATURE_MATCH(x) { }
 #endif /* < 3.4 */
